// @parity/hardhat-polkadot-resolc/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultBinaryResolcConfig = exports.defaultNpmResolcConfig = exports.COMPILER_REPOSITORY_API_URL = exports.COMPILER_REPOSITORY_URL = exports.TASK_COMPILE_SOLIDITY_LOG_DOWNLOAD_RESOLC_COMPILER_START = exports.TASK_COMPILE_SOLIDITY_RUN_RESOLC = exports.TASK_COMPILE_SOLIDITY_COMPILE_RESOLC = exports.TASK_COMPILE_SOLIDITY_GET_RESOLC_BUILD = exports.DEFAULT_TIMEOUT_MILISECONDS = exports.RESOLC_ARTIFACT_FORMAT_VERSION = exports.PLUGIN_NAME = void 0;
exports.PLUGIN_NAME = "hardhat-polkadot";
exports.RESOLC_ARTIFACT_FORMAT_VERSION = "hh-resolc-artifact-1";
exports.DEFAULT_TIMEOUT_MILISECONDS = 30000;
exports.TASK_COMPILE_SOLIDITY_GET_RESOLC_BUILD = "compile:solidity:resolc:get-build";
exports.TASK_COMPILE_SOLIDITY_COMPILE_RESOLC = "compile:solidity:resolc:compile";
exports.TASK_COMPILE_SOLIDITY_RUN_RESOLC = "compile:solidity:resolc:run";
exports.TASK_COMPILE_SOLIDITY_LOG_DOWNLOAD_RESOLC_COMPILER_START = "compile:solidity:log:download-resolc-compiler-start";
exports.COMPILER_REPOSITORY_URL = "https://github.com/paritytech/revive/releases/download/";
exports.COMPILER_REPOSITORY_API_URL = "https://api.github.com/repos/paritytech/revive/releases";
exports.defaultNpmResolcConfig = {
    version: "latest",
    compilerSource: "npm",
    settings: {},
};
exports.defaultBinaryResolcConfig = {
    version: "latest",
    compilerSource: "binary",
    settings: {
        optimizer: {
            enabled: true,
            runs: 200,
        },
    },
};
